action/penalization/allegation.go

```diff
@@ -0,0 +1,133 @@
+}
+
+func (r Allegation) Signers() []action.Address {
+	return []action.Address{r.ValidatorAddress.Bytes()}
+}
+
+func (r Allegation) Type() action.Type {
+	return action.ALLEGATION
+}
+
+func (r Allegation) Tags() kv.Pairs {
+	tags := make([]kv.Pair, 0)
+
+	tag := kv.Pair{
+		Key:   []byte("tx.type"),
+		Value: []byte(r.Type().String()),
+	}
+	tag2 := kv.Pair{
+		Key:   []byte("tx.validator"),
+		Value: r.ValidatorAddress.Bytes(),
+	}
+
+	tags = append(tags, tag, tag2)
+	return tags
+}
+
+var _ action.Tx = allegationTx{}
+
+type allegationTx struct{}
+
+func (atx allegationTx) Validate(ctx *action.Context, tx action.SignedTx) (bool, error) {
+	r := &Allegation{}
+	err := r.Unmarshal(tx.Data)
+	if err != nil {
+		return false, errors.Wrap(action.ErrWrongTxType, err.Error())
+	}
+	err = action.ValidateBasic(tx.RawBytes(), r.Signers(), tx.Signatures)
+	if err != nil {
+		return false, err
+	}
+
+	feeOpt, err := ctx.GovernanceStore.GetFeeOption()
+	if err != nil {
+		return false, gov.ErrGetFeeOptions
+	}
+	err = action.ValidateFee(feeOpt, tx.Fee)
+	if err != nil {
+		return false, err
+	}
+
+	if err := r.ValidatorAddress.Err(); err != nil {
+		return false, err
+	}
+
+	if err := r.MaliciousAddress.Err(); err != nil {
+		return false, err
+	}
+
+	if ctx.EvidenceStore.IsFrozenValidator(r.MaliciousAddress) {
+		return false, action.ErrFrozenValidator
+	}
+
+	if !ctx.EvidenceStore.IsActiveValidator(r.ValidatorAddress) {
+		return false, action.ErrNonActiveValidator
+	}
+
+	if r.ValidatorAddress.Equal(r.MaliciousAddress) {
+		return false, action.ErrInvalidAddress
+	}
+	return true, nil
+}
+
+func (atx allegationTx) ProcessCheck(ctx *action.Context, tx action.RawTx) (ok bool, result action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' transaction for ProcessCheck", tx)
+	ok, result = runAllegationTransaction(ctx, tx)
+	ctx.Logger.Debug("Result 'allegation' transaction for ProcessCheck", ok, result)
+	return
+}
+
+func (atx allegationTx) ProcessDeliver(ctx *action.Context, tx action.RawTx) (ok bool, result action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' transaction for ProcessDeliver", tx)
+	ok, result = runAllegationTransaction(ctx, tx)
+	ctx.Logger.Debug("Result 'allegation' transaction for ProcessDeliver", ok, result)
+	return
+}
+
+func (atx allegationTx) ProcessFee(ctx *action.Context, signedTx action.SignedTx, start action.Gas, size action.Gas) (bool, action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' Transaction for ProcessFee", signedTx)
+	return action.BasicFeeHandling(ctx, signedTx, start, size, 1)
+}
+
+func runAllegationTransaction(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	al := &Allegation{}
+	err := al.Unmarshal(tx.Data)
+	if err != nil {
+		return false, action.Response{Log: err.Error()}
+	}
+
+	err = ctx.EvidenceStore.PerformAllegation(al.ValidatorAddress, al.MaliciousAddress, al.BlockHeight, al.ProofMsg)
+	if err != nil {
+		return false, action.Response{Log: err.Error()}
```

Comment:
Can you keep the Log field the same pattern as we did in createProposal.go? The SDK side(not sure which side) need to parse the log string using regular expression or something.

Reply:
Update needed in all run****Transaction(...) functions. For Validate, you don't need to change.